README
-*- Text -*- is good.

  			    detachtty

detachtty lets you run interactive programs non-interactively, and
connect to them over the network when you do need to interact with them.

If you've used screen, it's a similar concept.  Compared to screen,
detachtty has

 * no capability to swap between different screens (it doesn't co-opt C-a)
 * no tty emulation (you can sanely use it in emacs comint modes)
 * no features (less to go wrong)
 * no configuration file
 * ability to spawn ssh to securely connect across the network

* What it's for - a likely usage scenario

I have a persistent server implemented in Lisp.  Usually it sits there
serving things, but sometimes (for example when a new version is
available and it needs to be upgraded) I need to connect to it and
type Lisp at it.

Why don't I use screen?  I could - I used to do exactly that - but
then I have to talk directly to the CMUCL repl, which means no command
line editing.  I want to use ILISP, but (1) screen's terminal handling
is likely to confuse Emacs, (2) screen isn't network-enabled and the
server machine doesn't have Emacs installed anyway.

* How it works:

detachtty has two parts: 

** detachtty  - the server

opens a pseudoterminal and starts your program on the slave end.  It
then copies between that tty and a unix-domain socket whose location
you specify when you start it

detachtty has options to fork into the background, to specify the
socket location, and to govern the placement of its log file and of
the dribble file from the application underneath it

** attachtty  - the client

When you want to connect to the process, attachtty attaches to the
unix-domain socket and copies input/output to and from you.  

If you have ssh installed and set up, you can also use attachtty
across the network.  This requires 

- that you also have attachtty installed on the machine that you are
  running detachtty on, and that it's somewhere on the path

It's often nice to be able to interrupt a process, especially when
it's stuck in some kind of loop.  attachtty catches SIGINT and sends
the character \003 into the pty master, which causes the controlled
process to get a SIGINT of its own.  This works at least for using
with ILISP and CMUCL.

To terminate attachtty, sending it SIGQUIT (usually from the
keyboard with CONTROL+\): it will detach from the unix-domain socket,
leaving the server process running. Sometimes you need to press it twice.

You can also suspend and resume attachtty as a regular Unix process
(usually from the keyboard with CONTROL+Z). The server process
will keep running while attachtty is suspended.

* Using it - an example

server$ detachtty --help
detachtty: unrecognized option `--help'
usage: 
 detachtty [--no-detach] [--dribble-file name] [--log-file name] \
   [--pid-file name] socket-path command [arg] [arg] [arg] ...

server$ detachtty --dribble-file /var/log/cliki-dribble --log-file \
  /var/log/detachtty.log --pid-file /var/run/cliki.pid \
  /var/run/cliki-socket /usr/bin/lisp -eval '(load-system :telentweb)' \
  -eval '(progn (telentweb::init) (telentweb::start))' \

client$ attachtty dan@server:/var/run/cliki-socket

** Signals and log files

When detachtty receives a SIGHUP, it closes and reopens its log and dribble
files, so it should play nice with whatever log rotation tools you already
use.  Note that the dribble file especially can grow large quite quickly  

The SIGHUP should be sent to detachtty, not to the child.  What the child
does with signals depends entirely on what the child is ...


* Security considerations

- detachtty performs no authentication of who's attached.  You choose
where the socket is created in the filesystem, so choose somewhere
that other people can't get to.

(detachtty creates its socket with mode 0700.  Some Unix systems -
Linux among them - honour file permissions on sockets, so only the
original user or root will be able to connect to it.  However, "many
BSD derived systems [...] ignore permissions for Unix sockets" so you
need to check if this will help you or not.)

- communication across the net is done using ssh.  If there is a
program on your computer called "ssh", then whatever level of security
it provides is available here

- no serious effort has been made to examine the attachtty and
detachtty code for buffer overruns etc.  If you are planning on using
these setuid or in some other untrusted-user situation, now is the
time to think of an alternative plan.


* Where to get it

detachtty development is hosted at <URL:https://github.com/cosmos72/detachtty>

Also, your distribution may (or may not) already contain it.
If it does, you will find that:

- on Debian, Ubuntu and other GNU/Linux systems that use ".deb" packages,
  "apt-get install detachtty" will download and install it
- on RedHat, Fedora, CentOS and other GNU/Linux systems that use ".rpm" packages,
  "yum install detachtty" will download and install it

Users of platforms that do not contain detachtty can download it from
<URL:https://github.com/cosmos72/detachtty/archive/master.tar.gz>
and compile and install it manually - see the INSTALL file inside the
downloaded package.


* Feedback, feature requests, patches

Use github issues at <URL:https://github.com/cosmos72/detachtty/issues>


ChangeLog
detachtty (11.0.0)

  * Added COPYING with GPLv2 license, updated README and INSTALL
  * Added explicit copyright statement to all sources and headers 
  * Added detachtty command-line option "--version"
  * removed debian/ folder, and moved ChangeLog from debian/ to top folder
  * Documented attachtty behaviour on SIGTSTP and CTRL+Z in man page

 -- Massimiliano Ghilardi <https://github.com/cosmos72> Sat, 14 Jan 1017 16:40:20 +0100

detachtty (10.5.0)

  * several improvements to attachtty:
  * propagate window size changes to process running inside detachtty
  * fixed tty initialization to correctly forward ^C ^Q ^R and ^S
  * added support for CTRL+Z
  * added support for timeout also on remote (ssh) connections
  * added clean exit upon end-of-file while reading from socket or standard input
  * when command is specified, terminate it with '\r' instead of '\n'
  * fixed signal handlers to set volatile flags, instead of calling fprintf() directly
  * simplified ssh remote connection logic, no longer requires ssh-askpass

  * and to detachtty:
  * updated man page detachtty.1
  * fixed several compile warnings and missing return values
  * improved logging messages
  * improved robustness of forking in background
  * do not pass pts_name[] to forkpty, as minimum buffer length cannot be determined

 -- Massimiliano Ghilardi <https://github.com/cosmos72> Sun, Apr 26 2015 19:52:16 +0200

detachtty (10) unstable; urgency=low

  * Minor manual page bugfix (Closes: #298877)
  * MacOS 10.3 doesn't need fink any more: note to that effect in the
    Makefile
  * More robust when the client goes away unexpectedly (ignore SIGPIPE,
    test for EPIPE)
  * Attempts to distinguish dead socket from earlier run and socket in
    use by simultaneous run 

 -- Daniel Barlow <dan@telent.net>  Thu, 31 Mar 2005 10:28:22 +0100

detachtty (9) unstable; urgency=low

  * Minor bug fix/incompatible change: reset the umask in the child process
    from 077 to whatever it had been in the parent environment
  * Bump debian standards version

 -- Daniel Barlow <dan@telent.net>  Sun, 14 Dec 2003 12:57:35 +0000

detachtty (8) unstable; urgency=low

  * Patch from Paul Lathrop to build on MacOS X with the libpoll fink package

 -- Daniel Barlow <dan@telent.net>  Mon,  4 Aug 2003 10:10:38 +0100

detachtty (7) unstable; urgency=low

  * Patch from Ng Pheng Siong to set VERASE2 to something harmless

 -- Daniel Barlow <dan@telent.net>  Thu, 23 Jan 2003 09:56:54 +0000

detachtty (6) unstable; urgency=low

  * Solaris port and assorted fixes (thanks to Matthias Koeppe and Utz-Uwe
    Haus)
  * Better error messages for bad command line arguments (Closes: #151156)
  * Manpage improvements
  * Probably survives a SIGHUP better if using relative log/dribble filenames

 -- Daniel Barlow <dan@telent.net>  Tue, 17 Sep 2002 22:06:05 +0100

detachtty (5) unstable; urgency=low

  * FreeBSD portability fixes
  * Includes text from strerror when reporting OS errors
  * Upload sponsored by Matthew Danish <mdanish@andrew.cmu.edu>

 -- Daniel Barlow <dan@telent.net>  Wed, 17 Oct 2001 10:59:03 +0100

detachtty (4) unstable; urgency=low

  * Dribble files now work correctly
  * Server process was dying in some circumstances when client went away
  * Upload sponsored by Matthew Danish <mdanish@andrew.cmu.edu>
  * More documentation

 -- Daniel Barlow <dan@telent.net>  Tue, 16 Oct 2001 19:54:51 +0100

detachtty (3) unstable; urgency=low

  * Man page fixes
  * Upload sponsored by Matthew Danish <mdanish@andrew.cmu.edu>
  * Closes: #110712 (ITP)
  * Re-upload due to initial rejection
    - ssh could not be Recommended if this package was to go into main.
    - Moved ssh to Suggests instead.

 -- Daniel Barlow <dan@telent.net>  Wed,  5 Sep 2001 13:32:53 -0400

detachtty (2) unstable; urgency=low

  * Fix various packaging infelicities
  * Removed #include reference to unused <popt.h>
  * Correct section number in manual page
  * Give intelligible error messages when invoked without args
  * Upload sponsored by Matthew Danish <mdanish@andrew.cmu.edu>
  * Closes: #110712 (ITP)

 -- Daniel Barlow <dan@telent.net>  Sat,  1 Sep 2001 00:51:21 +0100

detachtty (1) unstable; urgency=low

  * Initial Release.

 -- Daniel Barlow <dan@telent.net>  Thu, 31 May 2001 22:10:29 +0100

Local variables:
mode: debian-changelog
End:


config.h
/*
 * detachtty - run long-lived processes non interactively in the background,
 *             and reattach them to a terminal at later time
 *
 * Copyright (C) 2016-2017 Massimiliano Ghilardi
 * Copyright (C) 2001-2005 Daniel Barlow
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License
 *     as published by the Free Software Foundation; either version 2
 *     of the License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef DETACHTTY_CONFIG_H
#define DETACHTTY_CONFIG_H

#include <netdb.h>
#include <sys/poll.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <netinet/in.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>

#ifdef NEED_PTY_H
#include <pty.h>
#endif

#ifdef NEED_LIBUTIL_H
#include <libutil.h>
#endif

#ifdef NEED_STRINGS_H
#include <strings.h>
#endif

#ifdef NEED_LOCAL_FORKPTY_H
#include "forkpty.h"
#endif

#define DETACHTTY_VERSION_STR "11.0.0"

int copy_a_bit(int in_fd, int out_fd, int dribble_fd, char *message);
int copy_a_bit_with_log(int in_fd, int out_fd, int dribble_fd,
                        char * program, char *message);

int copy_a_bit_sendfd(int in_fd, int out_fd,
                      int dribble_fd, int send_fd, char *message);
int copy_a_bit_recvfd_with_log(int in_fd, int out_fd,
                               int dribble_fd, int * recv_fd,
                               char * program, char *message);

int output_buffer(int out_fd);
int output_buffer_sendfd(int out_fd, int send_fd);

extern FILE *log_fp;
int logprintf(char *progname, char *format, ...);
void bail(char *progname, char *format, ...);

#endif /* DETACHTTY_CONFIG_H */


errors.c
/*
 * detachtty - run long-lived processes non interactively in the background,
 *             and reattach them to a terminal at later time
 *
 * Copyright (C) 2016-2017 Massimiliano Ghilardi
 * Copyright (C) 2001-2005 Daniel Barlow
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License
 *     as published by the Free Software Foundation; either version 2
 *     of the License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdarg.h>
#include <signal.h>
#include <errno.h>
#include <time.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

FILE *log_fp = NULL;

int logprintf(char *progname, char *format,...) {
    va_list ap;
    int n;
    va_start(ap, format);
    fprintf(log_fp, ";;; %s: %ld: ", progname, time(NULL));
    n=vfprintf(log_fp, format, ap);
    va_end(ap);
    fputs("\r\n",log_fp);
    return n;
}

void bail(char *progname,char *format,...) {
    va_list ap;
    int err = errno;
    va_start(ap, format);
    fprintf(log_fp,";;; %s: %ld: FATAL ", progname, time(NULL));
    vfprintf(log_fp, format, ap);
    va_end(ap);
    /* use \r\n to avoid staircase effect */
    if (err > 0) 
	fprintf(log_fp, " (%s)\r\n", strerror(err));
    else
	fputs("\r\n",log_fp);

    kill(getpid(), SIGTERM);
}



forkpty.c
/*
 * detachtty - run long-lived processes non interactively in the background,
 *             and reattach them to a terminal at later time
 *
 * Copyright (C) 2016-2017 Massimiliano Ghilardi
 * Copyright (C) 2001-2005 Daniel Barlow
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License
 *     as published by the Free Software Foundation; either version 2
 *     of the License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <fcntl.h>
#include <stdlib.h>
#include <sys/termios.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/conf.h>
#include <stropts.h>

int forkpty (int *amaster, char *name, struct termios
	     *termp, struct winsize *winp)
{
    int fdm, fds;
    char *slavename;
    pid_t pid;
    fdm = open("/dev/ptmx", O_RDWR);  /* open master */
    grantpt(fdm);                     /* change permission of slave */
    unlockpt(fdm);                    /* unlock slave */
    slavename = ptsname(fdm);         /* get name of slave */
    if (name) strcpy(name, slavename);
    *amaster = fdm;
    if ((pid = fork()) < 0) {
	return pid; 		/* error */
    }
    else if (pid != 0) {		/* parent */
	return pid;
    }
    else {			/* child */
	pid_t pgid;
	/* create a new session */
	pgid = setsid();
	if (pgid == -1) {
	    perror("setsid failed");
	    return -1;
	}
	fds = open(slavename, O_RDWR);    /* open slave */
	ioctl(fds, I_PUSH, "ptem");       /* push ptem */
	ioctl(fds, I_PUSH, "ldterm");    /* push ldterm */
	dup2(fds, 0);
	dup2(fds, 1);
	dup2(fds, 2);
	ioctl(fds, TIOCSPGRP, &pgid);
	/* magic */
	if (termp)
	    ioctl(fds, TCSETS, termp);
	if (winp)
	    ioctl(fds, TIOCSWINSZ, winp);
        if (fds > 2)
            close(fds);
	return pid;
    }
}
      

int daemon(int nochdir, int noclose)
{
    int fd;

    switch (fork()) {
    case -1:
	return (-1);
    case 0:
	break;
    default:
	exit(0);
    }

    if (setsid() == -1)
	return (-1);

    if (!nochdir)
	chdir("/");

    if (!noclose && (fd = open("/dev/null", O_RDWR, 0)) != -1) {
	dup2(fd, 0);
	dup2(fd, 1);
	dup2(fd, 2);
	if (fd > 2)
	    close (fd);
    }
    return (0);
}



forkpty.h
/*
 * detachtty - run long-lived processes non interactively in the background,
 *             and reattach them to a terminal at later time
 *
 * Copyright (C) 2016-2017 Massimiliano Ghilardi
 * Copyright (C) 2001-2005 Daniel Barlow
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License
 *     as published by the Free Software Foundation; either version 2
 *     of the License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef DETACHTTY_FORKPTY_H
#define DETACHTTY_FORKPTY_H

#include <sys/termios.h>

int forkpty (int *amaster, char *name, struct termios
	     *termp, struct winsize *winp);

int daemon(int nochdir, int noclose);

#endif /* DETACHTTY_FORKPTY_H */


detachtty.c
/*
 * detachtty - run long-lived processes non interactively in the background,
 *             and reattach them to a terminal at later time
 *
 * Copyright (C) 2016-2017 Massimiliano Ghilardi
 * Copyright (C) 2001-2005 Daniel Barlow
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License
 *     as published by the Free Software Foundation; either version 2
 *     of the License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "config.h"

#include <errno.h>
#include <sys/stat.h>
#include <pty.h>
#include <fcntl.h>

#ifndef UNIX_PATH_MAX
# define UNIX_PATH_MAX    108
#endif

static char * const MY_NAME = "detachtty";

/*
  Create child process and establish the slave pseudo terminal as the
  child's controlling terminal.
  extern int forkpty __P ((int *__amaster, char *__name,
  struct termios *__termp, struct winsize *__winp));
  __amaster=fd of master pty
  __name = name of master pty, must be long enough.  how do we know?

  In the child, also dups the slave device to fds 0,1,2 and calls setsid()
  (which causes the right process group stuff to happen too)
*/

static int  parse_args(int argc, char *argv[], int * next_arg_ptr);
static void set_noecho(int fd);
static int  bind_socket_or_bail(char * socket_path);
static void usage(char *name,char *offending_option);
static void version(void);
static void init_signal_handlers(void);
static void process_accumulated_signals(void);
static void tidy_up_nicely(int signal);

/*
  1) fork and open pty pairs.  start child process on the slave end
  2) open network socket, listen to it - backlog 0
  (should be unix-domain socket)
  3) accept a single connection
*/

static char *socket_path=NULL;
static char *dribble_path=NULL;
static char *log_file_path=NULL;
static char *pid_file_path=NULL;
static int dribble_fd=-1;
static volatile int was_signaled = 0, was_sighupped = 0;

static void reopen_files(int signal) {
    if (log_file_path) {
        if (log_fp)
            fclose(log_fp);
        log_fp = fopen(log_file_path,"a");
        if (log_fp==NULL)
            perror("fopen log file");
        setvbuf(log_fp, NULL, _IONBF, 0); /* don't buffer */
        if (signal > 0)
            logprintf(MY_NAME, "Got signal %d, reopened log file \"%s\"",
                      signal, log_file_path);
    } else {
        log_fp = stderr;
    }
    if (dribble_path) {
        if (dribble_fd >= 0)
            close(dribble_fd);
        dribble_fd = open(dribble_path, O_WRONLY|O_CREAT|O_APPEND,0600);
        if (dribble_fd < 0)
            logprintf(MY_NAME,"Cannot open dribble file %s", dribble_path);
    }
}


static void usage(char *name,char *offending_option) {
    if(offending_option) 
        fprintf(stderr, "%s: unrecognized option `%s'\n",
                name,offending_option);
    else
        fprintf(stderr, "%s: unrecognized arguments\n",
                name);
    fprintf(stderr,"usage:\n"
            " %s [--version] [--no-detach] [--dribble-file name] [--log-file name] \\\n"
            "   [--pid-file name] socket-path /path/to/command [arg] [arg] [arg] ...\n",
            name);
}

static void version(void) {
    fprintf(stdout,"detachtty version %s\n%s", DETACHTTY_VERSION_STR,
            "Copyright (C) 2016-2017 Massimiliano Ghilardi\n"
            "Copyright (C) 2001-2005 Daniel Barlow\n"
            "License GPLv2+: GNU GPL version 2 or later <http://www.gnu.org/licenses/>.\n"
            "\n"
            "This is free software: you are free to change and redistribute it.\n"
            "There is NO WARRANTY, to the extent permitted by law.\n");
}


#define CLIENT_CONNECTED (sock>=0)

int main(int argc,char *argv[], char *envp[]) {
    struct pollfd ufds[3];
    struct termios my_termios;
    struct winsize my_winsize;
    int pty_master = -1, master_socket = -1, sock = -1, send_fd = -1;
    int next_arg = 0, pid = -1, detach_p = 1;
    mode_t old_umask;

    log_fp = stderr;	/* may be changed by parse_args */

    detach_p = parse_args(argc, argv, &next_arg);

    /* this assumes we're started from a shell.  would be smart to 
       default 80x24 or something if we can't do this */
    tcgetattr(0, &my_termios);
    ioctl(0, TIOCGWINSZ, &my_winsize);

    old_umask = umask(077);
    master_socket = bind_socket_or_bail(socket_path);
    umask(old_umask);

    if (listen(master_socket, 1) != 0)
        bail(MY_NAME,"listen");

    if (detach_p) {
        int dev_null;
        if (daemon(1, 1))
            bail(MY_NAME, "daemon");
        /* we leave stderr open - if the user really wanted his tty back,
           he'd have specified --log-file */
        close(0);
        close(1);
        dev_null = open("/dev/null",O_RDWR);
        dup2(dev_null, 0);
        dup2(dev_null, 1);
    }

    if (pid_file_path) {
        FILE *fp = fopen(pid_file_path, "w");
        if (fp != NULL) {
            fprintf(fp, "%d\n", (int)getpid());
            fclose(fp);
        }
    }
    reopen_files(0);

    setlinebuf(stdout);
    setlinebuf(stderr);

    pid = forkpty(&pty_master,NULL,&my_termios,&my_winsize);
    if (pid < 0) {
        /* error */
        perror("detach: Can't fork");
        exit(1);
    } else if (pid==0) {
        /* child */
        set_noecho(0);
        execve(argv[next_arg],&argv[next_arg],envp);
        bail(MY_NAME,"detach: exec failed");
    } else {
        /* parent */

        init_signal_handlers();

        logprintf(MY_NAME, "Successfully started"); 
        for (;;) {
            ufds[0].fd=pty_master; ufds[0].events=POLLIN|POLLHUP;
            ufds[1].fd=master_socket; ufds[1].events=POLLIN;
            ufds[2].fd=sock; ufds[2].events=POLLIN|POLLHUP;

            process_accumulated_signals();

            if (poll(ufds, CLIENT_CONNECTED ? 3 : 2, -1) == -1) {
                int err = errno;
                if (err != EINTR && err != EAGAIN)
                    logprintf(MY_NAME, "poll failed: %s", strerror(errno));
                continue;
            }
            process_accumulated_signals();

            if (ufds[0].revents & POLLIN) {
                if (copy_a_bit_sendfd(pty_master,sock,dribble_fd,send_fd,"copying from child") > 0)
                    send_fd = -1;
            }
            if (ufds[0].revents & POLLHUP) {
                logprintf(MY_NAME, "child terminated, exiting");
                if (sock>=0) { close(sock); sock=-1; }
                tidy_up_nicely(0);
            }

            if (ufds[1].revents & POLLIN) {
                struct sockaddr_un their_addr;
#ifdef __linux__
                socklen_t spare_integer = sizeof(their_addr);
#else
                int spare_integer = sizeof(their_addr);
#endif
                int new_sock = accept(master_socket, (struct sockaddr *) &their_addr,
                                      &spare_integer);
                if (new_sock >= 0)
                {
                    logprintf(MY_NAME,"accepted connection%s",
                              (CLIENT_CONNECTED ? " (and closing previous one)" : ""));
                    if (CLIENT_CONNECTED)
                        close(sock);
                    sock = new_sock;
                    send_fd = pty_master;
                    /* give them a copy of anything we read recently */
                    if (output_buffer_sendfd(sock, send_fd) > 0)
                        send_fd = -1;
                    continue;
                }
            }
            if (CLIENT_CONNECTED && (ufds[2].revents & POLLIN)) {
                if (copy_a_bit_with_log(sock, pty_master, dribble_fd,
                                        MY_NAME, "copying from socket, closing connection") == 0) {
                    /* end-of-file on socket */
                    if (sock >= 0) { close(sock); sock=-1; }
                }
            }
            if (CLIENT_CONNECTED && (ufds[2].revents & POLLHUP)) {
                logprintf(MY_NAME, "closed connection due to hangup");
                close(sock); sock=-1;
            }
        }
    }
    return 0;
}

static int parse_args(int argc, char *argv[], int * next_arg_ptr) {
    int next_arg, detach_p = 1;
    for (next_arg = 1; next_arg < argc; next_arg++) {
        if (!strcmp("--no-detach", argv[next_arg])) {
            detach_p = 0;
        }
        else if (!strcmp("--dribble-file", argv[next_arg])) {
            dribble_path = argv[++next_arg];
        }
        else if (!strcmp("--log-file", argv[next_arg])) {
            log_file_path = argv[++next_arg];
        }
        else if (!strcmp("--pid-file", argv[next_arg])) {
            pid_file_path = argv[++next_arg];
        }
        else if (!strcmp("--version", argv[next_arg])) {
            version();
            exit(0);
        }
        else if (!strncmp("--", argv[next_arg], 2)) {
            usage(argv[0], argv[next_arg]);
            exit(1);
        }
        else {
            break;
        }
    }
    if (next_arg >= argc - 1) {
        usage(argv[0], NULL);
        exit(1);
    }
    socket_path=argv[next_arg++];

    if (argv[next_arg][0] != '/') {
        logprintf(MY_NAME,"\"%s\" is not an absolute path", argv[next_arg]);
        bail(MY_NAME, "argument parsing");
    }
    *next_arg_ptr = next_arg;
    return detach_p;
}

static int bind_socket_or_bail(char * socket_path) {
    char pidbuf[80];
    struct sockaddr_un addr;
    FILE * fp = NULL;
    int master_socket, oldpid = -1;

    master_socket = socket(PF_UNIX, SOCK_STREAM, 0);
    if (master_socket < 0) {
        bail(MY_NAME, "socket");
    }
    addr.sun_family=AF_UNIX;
    strncpy(addr.sun_path, socket_path, UNIX_PATH_MAX);
    addr.sun_path[UNIX_PATH_MAX-1] = '\0';

    if (bind(master_socket, (const struct sockaddr *) &addr, sizeof(addr)) == 0) {
        return master_socket;
    }
    if (!pid_file_path) {
        goto no_pid;
    }
    fp = fopen(pid_file_path,"r");
    if (fp) {
        if (fgets(pidbuf, sizeof(pidbuf), fp)) {
            oldpid = atoi(pidbuf);
        }
        fclose(fp);
    }
    if (oldpid <= 0) {
        goto no_pid;
    }
    if (kill(oldpid,0) == 0 || errno != ESRCH) {
        logprintf(MY_NAME, "process %d for pid file \"%s\" is still running",
                  oldpid, pid_file_path);
        goto bind_failed;
    }
    /* remove socket_path and try again */
    if (unlink(pid_file_path) != 0) {
        goto bind_failed;
    }
    logprintf(MY_NAME, "found and removed stale socket \"%s\" from a previous run",
              socket_path);

    if (bind(master_socket, (const struct sockaddr *) &addr, sizeof(addr)) == 0) {
        return master_socket;
    } else {
        goto bind_failed;
    }

 no_pid:
    logprintf(MY_NAME, "Cannot create \"%s\": does it already exist from a previous run?",
              socket_path);
 bind_failed:
    if (master_socket >= 0)
        close(master_socket);

    bail(MY_NAME, "bind");
    return -1;
}


/* borrowed from APUE example code found at
   http://www.yendor.com/programming/unix/apue/pty/main.c
*/

/* turn off echo (for slave pty) */
static void set_noecho(int fd) {
#ifdef NEED_SET_NOECHO
    struct termios  stermios;

    if (tcgetattr(fd, &stermios) < 0)
        bail("detach child","tcgetattr error");

    stermios.c_lflag &= ~(ECHO | ECHOE | ECHOK | ECHONL);
    stermios.c_lflag|=ICANON;
    /* stermios.c_oflag &= ~(ONLCR); */
    /* would also turn off NL to CR/NL mapping on output */
    stermios.c_cc[VERASE]=0177;
#ifdef VERASE2
    stermios.c_cc[VERASE2]=0177;
#endif
    if (tcsetattr(fd, TCSANOW, &stermios) < 0)
        bail("detach child","tcsetattr error");
#else
    (void)fd;
#endif /* NEED_SET_NOECHO */
}

static void cleanup_signal_handler(int sig) {
    struct sigaction act;
    act.sa_handler = SIG_DFL;
    sigemptyset(&(act.sa_mask));
    act.sa_flags = 0;
    sigaction(sig, &act, 0);
}

static void fatal_signal_handler(int sig) {
    was_signaled = sig;
    cleanup_signal_handler(sig);
}

static void sighup_signal_handler(int sig) {
    was_sighupped = sig;
}

static void init_signal_handlers(void) {
    struct  sigaction act;
    int fatal_sig[] = {
        SIGHUP, SIGQUIT, SIGILL, SIGABRT, SIGBUS, SIGFPE, SIGSEGV,
        /*SIGPIPE,*/ SIGTERM, SIGSTKFLT, SIGCHLD, SIGXCPU, SIGXFSZ,
    };
    unsigned i;

    /* catch SIGCHLD, SIGQUIT, SIGTERM, SIGILL, SIGFPE... and exit */
    act.sa_handler = fatal_signal_handler;
    sigemptyset(&(act.sa_mask));
    act.sa_flags = SA_RESETHAND;
    for (i = 0; i < sizeof(fatal_sig)/sizeof(fatal_sig[0]); i++) {
        sigaction(fatal_sig[i],&act,0);
    }

    /* we can be HUPped */
    act.sa_handler = sighup_signal_handler;
    sigemptyset(&(act.sa_mask));
    act.sa_flags = 0;
    sigaction(SIGHUP, &act,0);

    /* ignore SIGPIPE */
    act.sa_handler = SIG_IGN;
    sigemptyset(&(act.sa_mask));
    act.sa_flags = 0;
    sigaction(SIGPIPE, &act,0);
}

static void process_accumulated_signals(void) {
    if (was_signaled) {
        tidy_up_nicely(was_signaled);
        was_signaled = 0;
    }
    if (was_sighupped) {
        reopen_files(was_sighupped);
        was_sighupped = 0;
    }
}

static void tidy_up_nicely(int signal) {
    if (signal)
        logprintf(MY_NAME, "got unexpected signal %d, exiting", signal);
    else
        logprintf(MY_NAME, "exiting", signal);
    if (unlink(socket_path))
        logprintf(MY_NAME, "error unlinking \"%s\": %s", socket_path, strerror(errno));
    if (pid_file_path && unlink(pid_file_path) != 0)
        logprintf(MY_NAME, "error unlinking \"%s\": %s", pid_file_path, strerror(errno));
    exit(signal);
}


/*
 * Local Variables:
 * c-basic-offset: 4
 * End:
 */
